File: backend/src/main/java/com/dominoes/lobby/entity/Lobby.java
package com.dominoes.lobby.entity;

import com.dominoes.lobby.domain.PieceEnum;
import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OrderColumn;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "lobby")
@Getter
@Setter
@NoArgsConstructor
public class Lobby {

    @Id
    private UUID id;

    @Column(nullable = false)
    private Integer size;

    @Column(nullable = false)
    private boolean inProgress = false;

    @Column(name = "current_player_id")
    private UUID currentPlayerId;

    @Column(name = "opening_piece")
    private PieceEnum openingPiece;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "lobby_boneyard", joinColumns = @JoinColumn(name = "lobby_id"))
    @Column(name = "piece", nullable = false)
    @OrderColumn(name = "piece_order")
    private List<PieceEnum> boneyard = new ArrayList<>();

    public Lobby(Integer size) {
        this.id = UUID.randomUUID();
        this.size = size;
    }
}

File: backend/internal/entity/lobby.go
package entity

import (
	"time"

	"github.com/dominoes/lobby/internal/domain"
	"github.com/google/uuid"
)

type Lobby struct {
	ID              uuid.UUID
	Size            int
	InProgress      bool
	CurrentPlayerID *uuid.UUID
	DrawnThisTurn   bool
	Boneyard        []domain.PieceEnum
	Table           []TablePiece
}

func NewLobby(size int) *Lobby {
	return &Lobby{
		ID:         uuid.New(),
		Size:       size,
		Boneyard:   []domain.PieceEnum{},
		Table:      []TablePiece{},
	}
}

type User struct {
	ID        uuid.UUID
	Nickname  string
	JoinedAt  time.Time
	LobbyID   uuid.UUID
	Hand      []domain.PieceEnum
}

func NewUser(nickname string, lobby *Lobby) *User {
	return &User{
		ID:       uuid.New(),
		Nickname: nickname,
		JoinedAt: time.Now(),
		LobbyID:  lobby.ID,
		Hand:     []domain.PieceEnum{},
	}
}
